### PR TITLE
Feature - Smart Contract

### PR DESCRIPTION
**Pull request by KiranHRubix**
This Pull Request is a master codebase that contains the individual code base and logic for :

1. Smart Contract Swagger -> PR closes #40 
2. Smart Contract Creation and Fetch -> PR closes #40 
3. Smart Contract Pub Sub -> PR closes #59 
4. TokenState Pinning and checking (txn security) -> PR closes #52
5. Smart Contract Deploy and Execute
6. Utility. methods for the smart contract

The Issues that will be closed up on this PR getting merged are as follows 
1. Issue #46 
2. Issue #56 
3. Issue #57

### CODE REVIEW

File: .vscode/launch.json
Comment (Allen-Iype) [inline .vscode/launch.json:1]: I think we should add this .vscode also to .gitignore

File: block/block.go [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+// 	 "3" : TokenOwner       : string
+// 	 "4" : GenesisBlock     : GenesisBlock
+//   "5" : TransInfo        : TransInfo
+//   "6" : SmartContract    : []byte
+//   "7" : QuorumSignature  : []string
+//   "8" : PledgeDetails    : map[string][]PledgeDetail
+//   "9" : SmartContractData : string
+//
+// }
+
```
Comment (Allen-Iype) [inline block/block.go:26]: The comment above indicates what each number in the tokenchain means. Right ? @KiranHRubix

File: .vscode/launch.json
Comment (KiranHRubix) [reply .vscode/launch.json:1]: the launch.json file is used to run automated test in debug mode. providing this as a sample would be beneficial right?

File: block/block.go (hunk 1 above)
Comment (KiranHRubix) [reply block/block.go:26]: yes that is correct

File: .vscode/launch.json
Comment (Allen-Iype) [reply .vscode/launch.json:1]: Yes, providing this as a sample is a good idea.

File: command/command.go [hunk 2]
```diff
@@ -125,70 +150,93 @@ var commandsHelp = []string{"To get tool version",
...
+	date               time.Time
+	grpcAddr           string
+	grpcPort           int
+	grpcSecure         bool
+	deployerAddr       string
+	binaryCodePath     string
+	rawCodePath        string
+	schemaFilePath     string
+	smartContractToken string
+	newContractBlock   string
```
Comment (vaishnav-ch) [inline command/command.go:235]: I think the `newContractBlock` flag is not needed as changes are fetched from IPFS. Can I remove the same @KiranHRubix ?

File: command/diagnostic.go [hunk 3]
```diff
@@ -168,3 +170,61 @@ func (cmd *Command) dumpTokenChain() {
...
+	// format willbe json object
+	/*
+		{
+			block_no
+			block_hash
+			smartcontract_hash
+		}
+	*/
+
+}
```
Comment (vaishnav-ch) [inline command/diagnostic.go:230]: Is this Command/function needed.

File: command/command.go (hunk 2 above)
Comment (vaishnav-ch) [inline command/command.go:235]: I think the `newContractBlock` flag is not needed as changes are fetched from IPFS. Can I remove the same @KiranHRubix ?

File: command/diagnostic.go (hunk 3 above)
Comment (vaishnav-ch) [inline command/diagnostic.go:230]: Is this Command/function needed?

File: core/core.go [hunk 4]
```diff
@@ -476,6 +493,7 @@ func (c *Core) RemoveWebReq(reqID string) *ensweb.Request {
 
 func (c *Core) SetupDID(reqID string, didStr string) (did.DIDCrypto, error) {
 	dt, err := c.w.GetDID(didStr)
+	c.log.Debug("dt is", "dt", dt)
```
Comment (vaishnav-ch) [inline core/core.go:496]: Are these debug logging still needed ?

**Review by vaishnav-ch — APPROVED** [review]

File: contract/contract.go [hunk 5]
```diff
@@ -306,6 +333,49 @@ func (c *Contract) GetTransTokenInfo() []TokenInfo {
 
 }
 
+func (c *Contract) GetCommitedTokensInfo() []TokenInfo {
```
Comment (harirubix) [inline contract/contract.go:336]: Similar to GetTransTokenInfo() . Can this function be moved to a generic utils ?

File: core/diagnostic.go [hunk 6]
```diff
@@ -30,3 +34,127 @@ func (c *Core) DumpTokenChain(dr *model.TCDumpRequest) *model.TCDumpReply {
 	ds.NextBlockID = nextID
 	return ds
 }
+
+func (c *Core) DumpSmartContractTokenChain(dr *model.TCDumpRequest) *model.TCDumpReply {
```
Comment (harirubix) [inline core/diagnostic.go:38]: Again , feel we can generalise This to DumpTokenChain() .

File: contract/contract.go (hunk 5 above)
Comment (KiranHRubix) [reply contract/contract.go:336]: TransTokenInfo info lie with in a separate nested block, where as commited tokens info is mostly present in gensys block, hence written as a separate function

File: core/diagnostic.go (hunk 6 above)
Comment (KiranHRubix) [reply core/diagnostic.go:38]: Yes i think this can be clubbed together

File: contract/contract.go (hunk 5 above)
Comment (harirubix) [reply contract/contract.go:336]: To be handled during [optimisation Milestone](https://github.com/rubixchain/rubixgoplatform/milestone/2)

**Review by harirubix — APPROVED** [review]

**Review by Allen-Cherian — APPROVED** [review]

**Pull request merged**